internal/sim/api.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sim...
+//
+// TODO(mwhittaker): Write comprehensive package documentation with examples.
+// We also probably want to put some of this documentation on the website, and
+// we might also want to write a blog.
+//
+// TODO(mwhittaker): Move things to the weavertest package.
+package sim
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ServiceWeaver/weaver/internal/reflection"
+	"github.com/ServiceWeaver/weaver/weavertest"
+)
+
+// A Generator[T] generates random values of type T.
+type Generator[T any] interface {
+	// Generate returns a randomly generated value of type T. While Generate is
+	// "random", it must be deterministic. That is, given the same instance of
+	// *rand.Rand, Generate must always return the same value.
+	Generate(*rand.Rand) T
+}
+
+// A Registrar is used to register stuff with a simulator.
+type Registrar interface {
+	// RegisterFake registers a fake implementation of a component.
+	RegisterFake(weavertest.FakeComponent)
+
+	// RegisterGenerators registers generators for a workload method, one
+	// generator per method argument. The number and type of the registered
+	// generators must match the method. For example, given the method:
+	//
+	//     Foo(context.Context, int, bool) error
+	//
+	// we must register a Generator[int] and a Generator[bool]:
+	//
+	//     var r Registrar = ...
+	//     var i Generator[int] = ...
+	//     var b Generator[bool] = ...
+	//     r.RegisterGenerators("Foo", i, b)
```

Comment:
What happens if a generator isn't specified for a method? Is the plan to default to integers{}, float64s{} etc?

Reply:
I'm not sure yet. We could use a set of defaults (like integers{}), but I also don't think it's too wild to require that the user register generators for every method. For now, I require it, but let's revisit later?

Reply:
SGTM